toolchain/driver/clang_runner.cpp

```diff
@@ -0,0 +1,152 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/driver/clang_runner.h"
+
+#include <algorithm>
+#include <memory>
+#include <numeric>
+#include <optional>
+
+#include "clang/Basic/Diagnostic.h"
+#include "clang/Basic/DiagnosticOptions.h"
+#include "clang/Driver/Compilation.h"
+#include "clang/Driver/Driver.h"
+#include "clang/Frontend/CompilerInvocation.h"
+#include "clang/Frontend/TextDiagnosticPrinter.h"
+#include "common/command_line.h"
+#include "common/vlog.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/ScopeExit.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/Program.h"
+#include "llvm/Support/VirtualFileSystem.h"
+#include "llvm/TargetParser/Host.h"
+
+namespace Carbon {
+
+static auto GetExecutablePath(llvm::StringRef exe_name) -> std::string {
+  // If the `exe_name` isn't already a valid path, look it up.
+  if (!llvm::sys::fs::exists(exe_name)) {
+    if (llvm::ErrorOr<std::string> path_result =
+            llvm::sys::findProgramByName(exe_name)) {
+      return *path_result;
+    }
+  }
+
+  return exe_name.str();
+}
+
+ClangRunner::ClangRunner(llvm::StringRef exe_name, llvm::StringRef target,
+                         llvm::raw_ostream* vlog_stream)
+    : exe_name_(exe_name),
+      exe_path_(GetExecutablePath(exe_name)),
+      target_(target),
+      vlog_stream_(vlog_stream),
+      diagnostic_ids_(new clang::DiagnosticIDs()) {}
+
+auto ClangRunner::Run(llvm::ArrayRef<llvm::StringRef> args) -> bool {
+  // TODO: Maybe handle response file expansion similar to the Clang CLI?
+
+  // If we have a verbose logging stream, and that stream is the same as
+  // `llvm::errs`, then add the `-v` flag so that the driver also prints verbose
+  // information.
+  bool inject_v_flag = vlog_stream_ == &llvm::errs();
+  llvm::ArrayRef<llvm::StringRef> maybe_v_flag;
+  if (inject_v_flag) {
+    maybe_v_flag = {"-v"};
+  }
+
+  CARBON_CHECK(!args.empty());
+  CARBON_VLOG() << "Running Clang driver with arguments: \n";
+
+  // Render the arguments into null-terminated C-strings for use by the Clang
+  // driver. Command lines can get quite long in build systems so this tries to
+  // minimize the memory allocation overhead.
+  int total_size = std::accumulate(
+      args.begin(), args.end(), 0,
+      [](int sum, llvm::StringRef arg) { return sum + arg.size(); });
+  // Factor in the executable name and all the null bytes.
+  total_size += exe_name_.size() + args.size() + 1;
+  if (inject_v_flag) {
+    total_size += strlen("-v") + 1;
+  }
```

Comment:
There were a couple things that I was wondering about this:
- Why std::accumulate instead of a range-based for loop? 
- Is the `+ args.size()` for performance? I was having trouble reading this, how significant would the trade-off be to do `+ 1`? (alternately, making the comment more specific to `args.size()` might help)
- Had you considered something like `maybe_v_flag[0].size()` instead of `strlen("-v")`?

But I'm wondering, had you considered the below loop (reusing `args_range` there):

```
  auto args_range = llvm::concat<const llvm::StringRef>(
      llvm::ArrayRef({exe_name_}), maybe_v_flag, args);
  int total_size = 0;
  for (llvm::StringRef arg : args_range) {
    total_size += arg.size() + 1;
  }
```

Reply:
Sorry, some of this complexity got added later on in the evolution of teh PR, I can try to simplify. I like the idea of extracting the entire range now that we have so many things in it.

My inclination is always to use an algorithm rather than a range based for loop here, but the code for that gets quite bad here so switched to a loop.